### DISCUSSION TITLE
Linux style commands, ANSI terminal control, and a fledgling text editor

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#9919] [Show and tell]
# REPL Buddy

"Let me kick it like it's 1986 now." -- _Portugal the Man_

Do you ever get nostalgic for the old days? When 64K was a lot of RAM, home computers booted to a BASIC prompt on a green screen, and logging onto the VAX cluster meant a trek across campus to the Engineering building?

Just me? No?

If not, I've been working on something called REPL Buddy you might like.

There's a few parts to it:

## command.py
```
from command import *
```

Now you can do shell-like things: `cat()`, `ls()`, `grep()`, etc., albeit with function parameters rather than command-line arguments. Fifteen functions modeled after popular Linux shell commands.

## ansi.py
```
from ansi import ANSI
terminal = ANSI()
```

Now you have control over cursor position, text color, inverse and underline. You can also read keypresses such as arrow keys, home, end, page up/down, F1 .. F4, CTRL+key combinations, etc. All from MicroPython.

>Note: You'll need to connect to the REPL with a terminal emulator. Thonny does not play well with ANSI.

## femto.py
```
from femto import Femto
editor = Femto('filename.txt')
```

Now you have a woefully incomplete, still in development, totally not ready for prime time, shamefully undocumented text editor in the style of GNU Nano (or University of Washington's Pico) implemented as a Python class.

But be warned, you can't actually edit anything at the moment, just read, write, and page through text files. I did mention it's still in development, didn't I?

## How to Install It
You can install REPL Buddy with mip. The package includes command.py and ansi.py. You'll have to get femto.py separately since it's still incomplete.

Windows
```
py -m mpremote mip install github:DavesCodeMusings/repl-buddy
```

Linux
```
mpremote mip install github:DavesCodeMusings/repl-buddy
```

## How to Use It
After importing, command.py functions can be used any time at the REPL prompt. Available functions are listed in the [project](https://github.com/DavesCodeMusings/repl-buddy/) [README](https://github.com/DavesCodeMusings/repl-buddy/blob/main/README.md)

For ansi.py, you'll need an ANSI terminal emulator. PuTTY works well on Windows. Not sure about Linux, but if you find something that works, let me know. The ANSI class is not well documented at the moment, other than code comments. But, if you've programmed with ncurses before, you'll probably catch on.

## Parting Thoughts
This whole thing is a bit outside the usual 'read a sensor; publish to mqtt; repeat' kind of MicroPython, but it's been an entertaining way to learn. I'm sharing because in this great big world, there might be someone else interested in this same kind of retro geekery. It really does take me back to my first computer and late night CompSci homework on VT100 terminals. And even though MicroPython is lightyears ahead of BASIC and Pascal, the same constraits of limited resources make all of this an interesting challenge.

Check it out and let me know what you think!

**Comment by scruss** [comment]
Neat! You might be duplicating some effort from [robert-hh/Micropython-Editor: Small on-board editor for PyBoard, WiPy, ESP8266, ESP32, RP2040, W60x, XBEE 3, Sipeed K210, as well as PyCom and Adafruit devices written in Python](https://github.com/robert-hh/Micropython-Editor), though.

(ah, connecting to the Vax cluster via an Atari ST terminal from an engineering lab, hoping that I could download my Archie ftp queue before my 512 KB — yes, kilobytes — disk quota kicked in for the day ... It was all a bit crap back then, wasn't it?)

**Reply by DavesCodeMusings** [reply]
Duplicating effort seems to be my modus operandi. I wrote a little Python Flask-type thing called Thimble only to find out about microdot after posting in this forum. Ah, well. The editor robert-hh built seems very robust compared to my attempt. I'll probably keep going though as I'm learning a lot in the process.

I don't recall my disk quota from back then, but I do remember using a hole punch to double-side some 5.25" floppies. We've come a long way.

**Comment by peterhinch** [comment]
Also perhaps duplicating [official upysh](https://github.com/micropython/micropython-lib/blob/038b4ac6572d0d5b4c18148dc7d7fdc026369fa4/micropython/upysh/upysh.py) a shell offering Linux-style commands.

**Reply by DavesCodeMusings** [reply]
True. Though I started this project mainly for my own learning. And to that end, it's served me well.
